tools/flaky-test-retryer/github_commenter.go

```diff
@@ -37,3 +53,107 @@ func NewGithubClient(githubAccount string) (*GithubClient, error) {
+	if err != nil {
+		return err
+	}
+	oldEntries, err := parseEntries(oldComment)
+	if err != nil {
+		return err
+	}
+	if _, ok := oldEntries[jd.JobName]; !ok {
+		oldEntries[jd.JobName] = 0
+	}
+	newComment, canRetry := buildNewComment(jd, oldEntries, outliers)
+	if !canRetry {
+		return fmt.Errorf("expended all %d retries", maxRetries)
+	}
+	if dryrun {
+		log.Printf("[dry run] Comment not updated. See it here:\n%s\n", newComment)
+		return nil
+	}
+	if oldComment != nil {
+		if err := gc.DeleteComment(jd.Refs[0].Org, jd.Refs[0].Repo, oldComment.GetID()); err != nil {
+			return err
+		}
+	}
+	_, err = gc.CreateComment(jd.Refs[0].Org, jd.Refs[0].Repo, jd.Refs[0].Pulls[0].Number, newComment)
+	return err
+}
+
+// getOldComment queries the GitHub PR specified and gets the comment made by us. If no comment
+// is found, we do not error, since we will be creating a new one anyways.
+func (gc *GithubClient) getOldComment(org, repo string, pull int) (*github.IssueComment, error) {
+	comments, err := gc.ListComments(org, repo, pull)
+	if err != nil {
+		return nil, err
+	}
+	user, err := gc.GetGithubUser()
+	if err != nil {
+		return nil, err
+	}
+	for _, comment := range comments {
+		if *comment.GetUser().Login == *user.Login {
+			return comment, nil
+		}
+	}
+	return nil, nil
+}
+
+// parseEntries collects retry information from the given comment, so we can reuse it in
+// a new comment.
+func parseEntries(comment *github.IssueComment) (map[string]int, error) {
+	entries := map[string]int{}
+	if comment == nil {
+		return entries, nil
+	}
+	re := regexp.MustCompile(`.* \| \d`)
+	entryStrings := re.FindAll([]byte(comment.GetBody()), -1)
+	for _, e := range entryStrings {
+		fields := strings.Split(string(e), " | ")
+		retry, err := strconv.Atoi(strings.TrimSuffix(fields[1], "/3"))
+		if err != nil {
+			return nil, err
+		}
+		entries[strings.Trim(fields[0], "`")] = retry
+	}
+	return entries, nil
+}
+
+// buildNewComment takes the old entry data, the job we are processing, and any outlying
+// non-flaky tests, building a comment body based on these parameters.
+func buildNewComment(jd *JobData, entries map[string]int, outliers []string) (string, bool) {
+	var cmd string
+	var entryString []string
+	if len(outliers) > 0 {
+		cmd = buildNoRetryString(jd.JobName, outliers)
+	} else {
+		cmd = buildRetryString(jd.JobName, entries)
+	}
+	for test, retry := range entries {
+		entryString = append(entryString, fmt.Sprintf("%s | %d/%d", test, retry, maxRetries))
+	}
+	return fmt.Sprintf(commentTemplate, strings.Join(entryString, "\n"), cmd), entries[jd.JobName] <= maxRetries
+}
+
+// buildRetryString increments the retry counter and generates a /test string if we have
+// more retries available.
+func buildRetryString(job string, entries map[string]int) string {
+	entries[job]++
+	if entries[job] <= maxRetries {
+		return fmt.Sprintf("Automatically retrying...\n/test %s", job)
```

Comment:
How about listing tests that are flaky here? This can make the logic of `buildRetryString` and `buildNoRetryString` share majority of the logic

Reply:
As it stands, I don't collect the flaky tests that affect this job, only the outliers, so this will require some more logic elsewhere. Do you think it would be helpful from a development perspective, seeing which flaky tests affected that job?

Reply:
No one probably would even pay attention to that, but from my point of view, it would be useful for checking if this tool works properly